src/System.Threading.Channels/tests/Stress.cs

```diff
@@ -0,0 +1,144 @@
+        private static readonly int MaxTaskCounts = Math.Max(2, Environment.ProcessorCount);
+
+        [OuterLoop]
+        [Theory]
+        [MemberData(nameof(TestData))]
+        public void RunInStressMode(Func<ChannelOptions, Channel<int>> channelCreator, ChannelOptions options)
+        {
+            Channel<int> channel = channelCreator(options);
+            ChannelReader<int> reader = channel.Reader;
+            ChannelWriter<int> writer = channel.Writer;
+            BoundedChannelOptions boundedOptions = options as BoundedChannelOptions;
+            bool shouldReadAllWrittenValues = boundedOptions == null || boundedOptions.FullMode == BoundedChannelFullMode.Wait;
+
+            List<Task> taskList = new List<Task>();
+
+            int readerTasksCount;
+            int writerTasksCount;
+
+            if (options.SingleReader)
+            {
+                readerTasksCount = 1;
+                writerTasksCount = options.SingleWriter ? 1 : MaxTaskCounts - 1;
+            }
+            else if (options.SingleWriter)
+            {
+                writerTasksCount = 1;
+                readerTasksCount = MaxTaskCounts - 1;
+            }
+            else
+            {
+                readerTasksCount = MaxTaskCounts / 2;
+                writerTasksCount = MaxTaskCounts - readerTasksCount;
+            }
+
+            int readCount = 0;
+
+            for (int i=0; i < readerTasksCount; i++)
+            {
+                taskList.Add(Task.Run(async delegate
+                {
+                    while (true)
+                    {
+                        try
+                        {
+                            if (!reader.TryRead(out int value))
+                            {
+                                value = await reader.ReadAsync();
+                            }
+                            Interlocked.Increment(ref readCount);
+                        }
+                        catch (ChannelClosedException)
+                        {
+                            break;
+                        }
+                    }
+                }));
+            }
+
+            int numberToWriteToQueue = -1;
+            int remainingWriters = writerTasksCount;
+
+            for (int i=0; i < writerTasksCount; i++)
+            {
+                taskList.Add(Task.Run(async delegate
+                {
+                    int num = Interlocked.Increment(ref numberToWriteToQueue);
+                    while (num < MaxNumberToWriteToChannel)
+                    {
+                        await writer.WriteAsync(num);
+                        num = Interlocked.Increment(ref numberToWriteToQueue);
+                    }
+
+                    if (Interlocked.Decrement(ref remainingWriters) == 0)
+                        writer.Complete();
+                }));
+            }
+
+            Task.WaitAll(taskList.ToArray());
+
+            Assert.True(readCount <= MaxNumberToWriteToChannel);
```

Comment:
Why wouldn't they be equal?

Reply:
Ah, nevermind, I see the shouldReadAllWrittenValues below.  In that case I'd suggest a small tweak:
```C#
if (shouldReadAllWrittenValues)
{
    Assert.Equal(MaxNumberToWriteToChannel, readCount);
}
else
{
    Assert.InRange(readCount, 0, MaxNumberToWriteToChannel);
}
```

Reply:
ok, I'll add this.